submariner/IPV6-datapath.md

```diff
@@ -0,0 +1,129 @@
+| dual-stack | dual-stack | V4,V6 |
+
+connectivity across the clusters should be supported for the following cases:
+
+* Pods(including host-networking) to Services.
+* Pods (including host-networking) to Pods.
+
+## Proposal
+
+Currently, Submariner fully supports IPV4 inter-cluster connectivity, including egress in-cluster routing to reach GW node, GlobalNet, and inter-cluster tunnels.
+
+The idea is to duplicate intra-cluster and inter-cluster connectivity components also for IPV6.
+
+The active Gateway Engine communicates with the central Broker to advertise its Endpoint and Cluster resources to the other clusters connected to the Broker, also ensuring that it is the sole Endpoint for its cluster.
+The Endpoint resource fields should include IP addresses according to the cluster’s networking configuration.
+For example for a dual-stack cluster HealthCheckIP, PrivateIP, PublicIP and Subnets should consist of both IPv4 and IPV6 addresses.  
+
+The Route Agent running in the cluster learns about the local Endpoint and remote Endpoints and setup the necessary V4,V6 infrastructure to route cross-cluster traffic from all nodes to the active Gateway Engine node. 
+
+The active Gateway Engine also establishes a watch on the Broker to learn about the active Endpoint and Cluster resources advertised by the other clusters. Once two clusters are aware of each other’s Endpoints, they can establish a secure tunnel/s based on the content of the remote Endpoint and local Endpoint through which traffic can be routed. A tunnel should be created only if the local Endpoint networking type matches remote Endpoint ipfamily.
+
+The next diagram illustrates Submariner’s datapath architecture for kube-proxy based CNIs:
+![non-ovnk-architecture](./images/dual-stack-arch-for-non-ovnk.png)
+
+With the proposed architecture, Submariner needs to establish both V4 and V6 intra-cluster egress routing to the GW node in the case of dual-stack.
+
+Pod IPV4 egress packets for CNI != OVNK and cable-driver=libreswan will be:
+![non-ovnk-ipv4-egress](./images/ipv4-non-ovnk-egress-packets.png)
+
+And Pod IPV6 egress packets for the same configuration will be:
+![non-ovnk-ipv6-egress](./images/ipv6-non-ovnk-egress-packets.png)
+
+**Note**: In future, we may optimize this architecture for a dual-stack case.
+For example by using only the intra-cluster V4 VxLAN to route V4 and V6 traffic to the GW.
+
+## Datapath breakdown
+
+### Gateway
+
+To support IPV6 the gateway should:
+
+* discover publicIP, privateIP, healthcheckIP and cluster’s subnets for each IP family.  
+  * Note: gateway should address corner cases related to this change, for example in dual-stack environment only the V4 publicIP address is successfully resolved.    
+* run NAT Discovery per IP family in remote Endpoint.
+* advertise in local Endpoint IP details based on cluster networking type, for example in duals-stack cluster both V4 Public IP and V6 Public IP should be advertised in Endpoint.   
```

Comment:
We will not be having any explicity field to specify the type of cluster (i.e., IPv4 only, IPv6 only or dualStack), but the presence of multiple addresses in the endpoint will be like a signal that remote cluster has more than one IP for an element and depending on the type of IPaddress it could be v4/v6/both. I hope this is correct?

Reply:
Yep, that's correct.